File: Data Structures & Algorithms/matchsticks-to-square/submission-1.ts
class Solution {
    /**
     * @param {number[]} matchsticks
     * @return {boolean}
     */
    makesquare(matchsticks: number[]): boolean {
        matchsticks.sort((a,b)=> a-b);
        const total = matchsticks.reduce((acc, i)=> acc + i, 0)
        let rem = total%4;
        if(rem!==0 || matchsticks.length < 4){
            return false;
        }
        const side = total/4;
        let maxNum = Math.max(...matchsticks);
        if(maxNum > side){
            return false;
        }
        let res = false;
        // let count = 0;
        // let maskArray = new Array(matchsticks.length).fill(false);
        function dfs(t: number, b: number, l: number, r: number, i: number){
            // console.log("hell check", t,b,l,r,i)
            if(t===side && b===side && l===side && r === side){
                res = true;
                return
            }
            if(res){
                return
            }
            if(i >=matchsticks.length){
                return
            }
            
            if(t+matchsticks[i]<=side){
                dfs(t+matchsticks[i], b,l,r,i+1)
            }
            if(b+matchsticks[i]<=side){
                dfs(t, b +matchsticks[i],l,r,i+1)
            }
            if(l+matchsticks[i]<=side){
                dfs(t, b,l+matchsticks[i],r,i+1)
            }
            if(r+matchsticks[i]<=side){
                dfs(t, b,l,r+matchsticks[i],i+1)
            }
                

        }
        dfs(0,0,0,0, 0)
        return res
    }
}

File: Data Structures & Algorithms/longest-common-subsequence/submission-1.py
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        r = len(text1)
        c = len(text2)
        dp = {}

        def dfs(i,j):
            key = f"{i},{j}"
            if key in dp:
                return dp[key]
            if i >= r or j >= c:
                return 0
            
            if text1[i] == text2[j]:
                dp[key] = 1 + dfs(i+1, j+1)
            else:
                dp[key] = max(dfs(i,j+1), dfs(i+1, j))
            return dp[key]
        
        return dfs(0,0)